### PR TITLE
Log a warning when the BatchSpanProcessor queue is full.

### LINKED ISSUE #1870: Log a warning when BatchSpanProcessor queue is full and span is dropped
**Is your feature request related to a problem?**
Yes, the problem is that when the `BatchSpanProcessor` queue is full, spans are silently dropped. This may result in incomplete traces downstream. As lost spans can have many reasons, it is hard to pinpoint the problem if it originates from the BSP queue being full.

**Describe the solution you'd like**
It would be nice if a log message is emitted to notify the user a span got dropped because the queue was full. This will allow them to understand the problem and mitigate it quickly by tweaking their BSP queue size.

**Describe alternatives you've considered**
Making the queue blocking, but this is not allowed by the specification.

### PR DESCRIPTION
**Pull request by johanpel**
Fixes #1870 

## Changes

Adds a warning when a span is dropped in case the `BatchSpanProcessor` queue is full.

* [x] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [n/a] Changes in public API reviewed

### CODE REVIEW

File: sdk/test/trace/batch_span_processor_test.cc [hunk 1]
```diff
@@ -211,6 +234,18 @@ TEST_F(BatchSpanProcessorTestPeer, TestManySpansLoss)
 
   // Span should be exported by now
   EXPECT_GE(max_queue_size, spans_received->size());
+
+  // At least one log message warning about the dropped span should have been emitted by now.
+  auto &vec = dynamic_cast<MockLogHandler *>(log_handler.get())->messages;
```
Comment (owent) [inline sdk/test/trace/batch_span_processor_test.cc:239]: Should we use `static_cast` here?Sometimes we will compile otel-cpp without RTTI.(With `-fno-rtti` or `/GR-`)

Comment (johanpel) [reply sdk/test/trace/batch_span_processor_test.cc]: Thanks for pointing this out, CI was also failing because of this.

**Comment by johanpel** [comment]
> Thanks @johanpel for the PR.
> 
> FYI, there is a delay in running CI, because we have to manually allow it after a new commit, due to github rules for first time contributors.
> 
> I see the CLA is signed already, which is great.
> 
> Once this PR is merged, the process will be much easier ;-)

Makes sense. I've added a few commits to hopefully fix the CI issues. Most were related to the timing around the `BatchSpanProcessor`. With Valgrind/ASAN/TSAN the `BatchSpanProcessor` still seems to be able to keep the queue non-full in the `BatchSpanProcessorTestPeer/TestManySpansLoss` test. 

I've reduced the queue size. Hope that gives more consistent results.

**Comment by johanpel** [comment]
It was still a bit too flaky, added a commit to only check logs if anything is actually dropped.

**Review by owent — APPROVED** [review]
LGTM

**Review by marcalff — APPROVED** [review]
LGTM, thanks for the fix.

**Pull request merged**